### PR TITLE
Simple mechanism for service dependency injection

### PR DESCRIPTION
**Pull request by joao-borges**
Creating a simple dependency injection mechanism.
This is to avoid declaring a new entity in every registry that an implementation is needed.

This is achieved mixing functionality from three libraries:
- JSR-330: javax.inject basic API
- org.reflections: simple classloader scanner (https://github.com/ronmamo/reflections)
- com.google.inject:guice: implementation for JSR-330 and injection configuration

From now on, to add support for a new entity, only creation of classes and its implementations are needed. No more changes on registries and other implementations.

- Entity: annotate with `@Named` and specify entity type
- Linker (EntitiesLinker), Writer (EntityWriter) and Loader (EntityLoader): implement the interface, annotate class with `@Singleton`, create constructor with required dependencies and annotate with `@Inject` and add implementation.

If any new type of bean is needed, changes into `ExportPluginModule` will be necessary.

### CODE REVIEW

File: src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/bundle/entity/EntityUtils.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * of the MIT license.  See the LICENSE file for details.
+ */
+
+package com.ca.apim.gateway.cagatewayexport.tasks.explode.bundle.entity;
+
+import com.ca.apim.gateway.cagatewayexport.tasks.explode.bundle.Entity;
+
+import javax.inject.Named;
+
+public class EntityUtils {
```
Comment (apigatewayopensource) [inline src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/bundle/entity/EntityUtils.java]: Code Smell Code Smell: Add a private constructor to hide the implicit public one. ([squid:S1118](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AS1118))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-export-plugin&pullRequest=21&issues=AWWNK2Xo3Qu1svVwpGPv&open=AWWNK2Xo3Qu1svVwpGPv)

[ISSUE_KEY]: AWWNK2Xo3Qu1svVwpGPv

File: src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/bundle/BundleBuilder.java [hunk 2]
```diff
@@ -9,10 +9,12 @@
 import com.ca.apim.gateway.cagatewayexport.tasks.explode.bundle.entity.*;
 import com.ca.apim.gateway.cagatewayexport.tasks.explode.loader.EntityLoader;
 import com.ca.apim.gateway.cagatewayexport.tasks.explode.loader.EntityLoaderRegistry;
+import com.ca.apim.gateway.cagatewayexport.util.injection.ExportPluginModule;
```
Comment (apigatewayopensource) [inline src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/bundle/BundleBuilder.java]: Code Smell Code Smell: Remove this unused import 'com.ca.apim.gateway.cagatewayexport.util.injection.ExportPluginModule'. ([squid:UselessImportCheck](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-export-plugin&pullRequest=21&issues=AWWNK2Wg3Qu1svVwpGPt&open=AWWNK2Wg3Qu1svVwpGPt)

[ISSUE_KEY]: AWWNK2Wg3Qu1svVwpGPt

File: src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/loader/EntityLoaderRegistry.java [hunk 3]
```diff
@@ -6,25 +6,37 @@
...
+import com.google.common.annotations.VisibleForTesting;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
```
Comment (apigatewayopensource) [inline src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/loader/EntityLoaderRegistry.java]: Code Smell Code Smell: Remove this unused import 'java.util.stream.Collectors'. ([squid:UselessImportCheck](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-export-plugin&pullRequest=21&issues=AWWNK2cz3Qu1svVwpGPy&open=AWWNK2cz3Qu1svVwpGPy)

[ISSUE_KEY]: AWWNK2cz3Qu1svVwpGPy

File: src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/loader/EntityLoaderRegistry.java [hunk 4]
```diff
@@ -6,25 +6,37 @@
...
 
+import com.google.common.annotations.VisibleForTesting;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
```
Comment (apigatewayopensource) [inline src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/loader/EntityLoaderRegistry.java]: Code Smell Code Smell: Remove this unused import 'java.util.function.Function'. ([squid:UselessImportCheck](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-export-plugin&pullRequest=21&issues=AWWNK2cz3Qu1svVwpGPx&open=AWWNK2cz3Qu1svVwpGPx)

[ISSUE_KEY]: AWWNK2cz3Qu1svVwpGPx

File: src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/loader/EntityLoaderRegistry.java [hunk 5]
```diff
@@ -6,25 +6,37 @@
 
 package com.ca.apim.gateway.cagatewayexport.tasks.explode.loader;
 
+import com.google.common.annotations.VisibleForTesting;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Collections;
 import java.util.HashMap;
```
Comment (apigatewayopensource) [inline src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/loader/EntityLoaderRegistry.java]: Code Smell Code Smell: Remove this unused import 'java.util.HashMap'. ([squid:UselessImportCheck](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-export-plugin&pullRequest=21&issues=AWWNK2cz3Qu1svVwpGPw&open=AWWNK2cz3Qu1svVwpGPw)

[ISSUE_KEY]: AWWNK2cz3Qu1svVwpGPw

File: src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/loader/EntityLoaderRegistry.java [hunk 6]
```diff
@@ -6,25 +6,37 @@
 
 package com.ca.apim.gateway.cagatewayexport.tasks.explode.loader;
 
+import com.google.common.annotations.VisibleForTesting;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Collections;
```
Comment (apigatewayopensource) [inline src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/loader/EntityLoaderRegistry.java]: Code Smell Code Smell: Remove this unused import 'java.util.Collections'. ([squid:UselessImportCheck](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-export-plugin&pullRequest=21&issues=AWWNK2cz3Qu1svVwpGPz&open=AWWNK2cz3Qu1svVwpGPz)

[ISSUE_KEY]: AWWNK2cz3Qu1svVwpGPz

File: src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/bundle/entity/EntityTypeRegistry.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ *
+ */
+public class EntityTypeRegistry {
+
+    private final Map<String, Class<? extends Entity>> entityTypeMap;
+
+    @Inject
+    public EntityTypeRegistry(final Reflections reflections) {
+        Map<String, Class<? extends Entity>> entityTypeMap = new HashMap<>();
```
Comment (apigatewayopensource) [inline src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/bundle/entity/EntityTypeRegistry.java]: Code Smell Code Smell: Rename "entityTypeMap" which hides the field declared at line 25. ([squid:HiddenFieldCheck](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AHiddenFieldCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-export-plugin&pullRequest=21&issues=AWWNK2Xf3Qu1svVwpGPu&open=AWWNK2Xf3Qu1svVwpGPu)

[ISSUE_KEY]: AWWNK2Xf3Qu1svVwpGPu

File: src/main/java/com/ca/apim/gateway/cagatewayexport/util/injection/ExportPluginModule.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        bind(ExplodeBundle.class);
+    }
+
+    public static Injector getInjector() {
+        return ofNullable(injector).orElseGet(ExportPluginModule::create);
+    }
+
+    @VisibleForTesting
+    static Injector create() {
+        return injector = createInjector(new ExportPluginModule());
```
Comment (apigatewayopensource) [inline src/main/java/com/ca/apim/gateway/cagatewayexport/util/injection/ExportPluginModule.java]: Code Smell Code Smell: Extract the assignment out of this expression. ([squid:AssignmentInSubExpressionCheck](https://sonarcloud.io/organizations/ca-api-gateway/rules#rule_key=squid%3AAssignmentInSubExpressionCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=com.ca.apim.gateway%3Agateway-export-plugin&pullRequest=21&issues=AWWNK2ie3Qu1svVwpGP1&open=AWWNK2ie3Qu1svVwpGP1)

[ISSUE_KEY]: AWWNK2ie3Qu1svVwpGP1

File: src/main/java/com/ca/apim/gateway/cagatewayexport/util/injection/ExportPluginModule.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+        // bind the entity writers
+        Multibinder<EntityWriter> writersBinder = newSetBinder(binder(), EntityWriter.class);
+        reflections.getSubTypesOf(EntityWriter.class).forEach(l -> writersBinder.addBinding().to(l));
+        bind(EntityWriterRegistry.class);
+
+        // bind the entity type registry
+        bind(EntityTypeRegistry.class);
+
+        // bind the explode task impementation
```
Comment (yves-chan) [inline src/main/java/com/ca/apim/gateway/cagatewayexport/util/injection/ExportPluginModule.java]: small typo here in implementation

File: src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/bundle/entity/EntityUtils.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    /**
+     * @param entityClass entity class
+     * @param <E> entity type
+     * @return the entity type defined by annotation {@link Named} on each entity class
+     */
+    public static <E extends Entity> String getEntityType(Class<E> entityClass) {
+        Named named = entityClass.getAnnotation(Named.class);
+        return named != null ? named.value() : null;
```
[the remark this answers is not in the document]
Comment (joao-borges) [reply src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/bundle/entity/EntityUtils.java]: Yes, that's correct, named.value() is never null. `Named` may be null if someone forgot to add it to the entity implementation.

[the remark this answers is not in the document]
Comment (yves-chan) [reply src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/bundle/entity/EntityUtils.java]: Understood :+1:

**Pull request merged**